Data/HashMap/Internal/Array.hs

```diff
@@ -204,6 +209,23 @@ new _n@(I# n#) b =
 new_ :: Int -> ST s (MArray s a)
 new_ n = new n undefinedElem
 
+-- when shrinkSmallMutableArray# is available, the returned array is the same as the array given, as it is shrunk in place
+-- otherwise a copy is made
+shrink :: MArray s a -> Int -> ST s (MArray s a)
+#if MIN_VERSION_GLASGOW_HASKELL(8, 10, 7, 0)
+shrink mary _n@(I# n#) =
+  CHECK_GE("shrink", _n, (0 :: Int))
+  CHECK_LE("shrink", _n, (lengthM mary))
+  ST $ \s -> case Exts.shrinkSmallMutableArray# (unMArray mary) n# s of
+    s' -> (# s', mary #)
+#else
+shrink mary n = do
+  mary' <- new_ n
+  copyM mary 0 mary' 0 n
+  pure mary'
+#endif 
```

Comment:
I really think you should use `cloneSmallMutableArray#` here to avoid filling the new array with errors first.